[Perl] distributed training fixes and examples

tlby opened the pull request:
Gets the examples from /how_to/multi_devices.html such as
```
( cd example/image-classification &&
    python ../../tools/launch.py -n 2 --launcher ssh -H ~/hosts
        ./train_mnist.pl --network lenet --kv-store dist_sync )
```
to work in Perl.

tlby committed:
[Perl] bugfixes for distributed training using tools/launch.py

sergeykolychev commented on perl-package/AI-MXNet/lib/AI/MXNet/KVStore.pm:
```diff
@@ -28,35 +28,73 @@ use AI::MXNet::Function::Parameters;
 has '_updater_func' => (is => 'rw', isa => 'CodeRef');
 
-sub DEMOLISH
+method DEMOLISH
```
this should remain 'sub', method has to be defined as method(@args) and we do not use these args, so sub is preferred.

sergeykolychev commented on perl-package/AI-MXNet/lib/AI/MXNet/KVStore.pm:
```diff
@@ -72,30 +110,26 @@ sub DEMOLISH
     >>> # init a list of key-value pairs
     >>> $keys = [5, 7, 9]
-    >>> $kv->init(keys, [map { mx->nd->ones($shape) } 0..@$keys-1])
+    >>> $kv->init($keys, [map { mx->nd->ones($shape) } 1..@$keys])
```
-1 should remain

sergeykolychev commented on perl-package/AI-MXNet/lib/AI/MXNet/KVStore.pm:
```diff
@@ -127,15 +161,15 @@ method init(
-    >>> $kv->push($keys, [map { mx->nd->ones($shape) } 0..@$keys-1)
-    >>> $b = [map { mx->nd->zeros(shape) } 0..@$keys-1]
+    >>> $kv->push($keys, [map { mx->nd->ones($shape) } @$keys])
+    >>> $b = [map { mx->nd->zeros($shape) } @$keys]
```
the same -1 here

sergeykolychev commented on perl-package/AI-MXNet/lib/AI/MXNet/KVStore.pm:
```diff
@@ -200,13 +233,13 @@ method push(
     >>> # On single device
     >>> $keys = [5, 7, 9]
-    >>> $b = [map { mx->nd->zeros($shape) } 0..@$keys-1]
+    >>> $b = [map { mx->nd->zeros($shape) } @$keys]
```
-1

sergeykolychev commented on perl-package/AI-MXNet/lib/AI/MXNet/KVStore.pm:
```diff
@@ -200,13 +233,13 @@ method push(
         [ 2.  2.  2.]]
     >>> # On multiple devices
-    >>> $b = [map { [map { mx->nd->ones($shape, ctx => $_) } @$gpus ] } 0..@$keys-1]
+    >>> $b = [map { [ map { mx->nd->ones($shape, ctx => $_) } @$gpus] } @$keys]
```
-1

sergeykolychev commented on perl-package/AI-MXNet/lib/AI/MXNet/KVStore.pm:
```diff
@@ -332,21 +373,16 @@ method save_optimizer_states(Str $fname)
-    close(F);
-    $self->_updater->set_states($data);
+    assert(defined($self->_updater), "Cannot save states for distributed training");
+    $self->_updater->set_states(join '', IO::File->new($fname, 'rb').getlines());
```
->getlines()

tlby replied:
thanks, man.

sergeykolychev commented:
@tlby nice work as usual, just please fix these small things and submit to dmlc/mxnet, no need for this code to wait until the 'gluon' completion.

sergeykolychev committed:
providing explicit null pointer for the provided_arg_stypes. (#7791)

eric-haibin-lin committed:
fix wrong dist-kvstore push/pull/rsp_pull (#7762)

formath committed:
rm duplicated and unused code (#7764)

* rm not use variables

* rm duplicated and unused code

* bug

sergeykolychev commented:
@tlby 
this example probably should be moved to AI-MXNet/examples/distributed/ or if left in examples put into a dir that will make it clear that the code is in perl.

tlby commented:
There is already Python and R code in example/image-classification, you'd prefer the Perl separated?

sergeykolychev commented:
@tlby 
was not aware of .R files being there, I suppose we can add to that dir then, however keeping perl examples in AI-MXNet/examples dir has an added value of having them visible via metacpan
I'll defer to your judgement here.

ptrendx committed:
Fix Moderngpu usages in MXNet for CUDA 9 (#7789)

* Modify ModernGPU for CUDA 9

* Remove unused shfl_up that triggered compiler warning

kevinthesun committed:
Website redesign (#7787)

* Add background image

* Home page redesign

* Implement redesign for home page

* More features

* Address comments

* More fix

* Fix link

* Fix version build

* Move background image to web-data repository

* Small fix

* Dummy commit to retrigger test

* Move resources to web data

ptrendx committed:
Remove warnings about not used variable when compiling with CUDA 8 (#7769)

* Remove warnings about not used variable with CUDA 8

* Restart CI?

* Update depthwise_convolution_tf.cuh

* Remove unnecessary macro

* Restart CI

szha committed:
fix empty prefix case for namescope (#7794)

eric-haibin-lin committed:
Python symbol/ndarray api re-org (#7712)

* add ndarray folder

* add sparse.md draft

* update sparse.md

* add contrib, linalg, random to ndarray.md

* add contrib, linalg random to symbol

* add sparse.md

* update notes and op docs for sparse

* remove unused files

* add back random & la

* fix inherited methods for sparse ndarray

* remove speical members

* fix lint

* add contrib

* add random.md

tlby commented:
reducing the changes any further either result in distributed training crashes or failures to converge.

sergeykolychev commented on perl-package/AI-MXNet/lib/AI/MXNet/KVStore.pm:
```diff
@@ -244,15 +244,14 @@ method pull(
     }
     else
     {
-        $self->_updater(AI::MXNet::Optimizer->get_updater($optimizer));
```
will it work if this line left as is and 254 is $self->_set_updater($self->_updater) ? otherwise it seems like ->_updater is left undefined.
It's very weird that this is needed, the only thing this patch adds is the removal of the outer sub around the call to the updater. Do we really have stumbled upon proper bug in perl itself ?

sergeykolychev replied:
yes, ->_updater is used in this file in other subs, so it should be saved in the accessor

cjolivier01 committed:
Fix GPU/CUDNN memory allocation failure issues (#7795)

sergeykolychev commented on perl-package/AI-MXNet/lib/AI/MXNet/KVStore.pm:
```diff
@@ -371,7 +370,7 @@ method load_optimizer_states(Str $fname)
 =cut
 
-method _set_updater(CodeRef $updater_func)
+method _set_updater(AI::MXNet::Updater $updater_func)
```
tests assume that this function can accept the CodeRef as well, please make it Updater $updater_func
and apply this patch
```
diff --git a/perl-package/AI-MXNet/lib/AI/MXNet/Types.pm b/perl-package/AI-MXNet/lib/AI/MXNet/Types.pm
index b4ec7e9..4d801ed 100644
--- a/perl-package/AI-MXNet/lib/AI/MXNet/Types.pm
+++ b/perl-package/AI-MXNet/lib/AI/MXNet/Types.pm
@@ -37,6 +37,7 @@ class_type 'AI::MXNet::Optimizer';
 class_type 'AI::MXNet::Initializer';
 class_type 'AI::MXNet::InitDesc';
 class_type 'AI::MXNet::IRHeader';
+class_type 'AI::MXNet::Updater';
 subtype "AcceptableInput" => as "Num|PDL|PDL::Matrix|AI::MXNet::NDArray|AI::MXNet::NDArray::Slice|ArrayRef";
 subtype "Index"           => as "Int";
 subtype "DimSize"         => as "Int" => where { $_ >= 0 };
@@ -57,6 +58,7 @@ subtype "Callback"        => as "CodeRef|ArrayRef[Coderef]|AI::MXNet::Callback|A
 subtype "EvalMetric"      => as "AI::MXNet::EvalMetric|Str|CodeRef";
 subtype "Optimizer"       => as "AI::MXNet::Optimizer|Str";
 subtype "Initializer"     => as "AI::MXNet::Initializer|Str";
+subtype "Updater"         => as "AI::MXNet::Updater|CodeRef";
 subtype "Activation"      => as "AI::MXNet::Symbol|Str|CodeRef";
 subtype "SymbolOrArrayOfSymbols" => as "AI::MXNet::Symbol|ArrayRef[AI::MXNet::Symbol]";
 subtype "NameShapeOrDataDesc" => as "NameShape|AI::MXNet::DataDesc";
```

sergeykolychev commented on perl-package/AI-MXNet/lib/AI/MXNet/Module/Base.pm:
```diff
@@ -484,7 +484,7 @@ method fit(
     Maybe[Callback]|ArrayRef[Callback] :$epoch_end_callback=,
     Maybe[Callback]|ArrayRef[Callback] :$batch_end_callback=,
-    Str                                :$kvstore='local',
+    Str|AI::MXNet::KVStore             :$kvstore='local',
```
let's make it KVStore :$kvstore,
Below is full patch that should take care of the constraints for updater and KVStore.
```
developer@devbox:~/mxnet/perl-package/AI-MXNet/lib/AI/MXNet$ git diff Types.pm |less
diff --git a/perl-package/AI-MXNet/lib/AI/MXNet/Types.pm b/perl-package/AI-MXNet/lib/AI/MXNet/Types.pm
index b4ec7e9..27dd013 100644
--- a/perl-package/AI-MXNet/lib/AI/MXNet/Types.pm
+++ b/perl-package/AI-MXNet/lib/AI/MXNet/Types.pm
@@ -37,6 +37,8 @@ class_type 'AI::MXNet::Optimizer';
 class_type 'AI::MXNet::Initializer';
 class_type 'AI::MXNet::InitDesc';
 class_type 'AI::MXNet::IRHeader';
+class_type 'AI::MXNet::Updater';
+class_type 'AI::MXNet::KVStore';
 subtype "AcceptableInput" => as "Num|PDL|PDL::Matrix|AI::MXNet::NDArray|AI::MXNet::NDArray::Slice|ArrayRef";
 subtype "Index"           => as "Int";
 subtype "DimSize"         => as "Int" => where { $_ >= 0 };
@@ -48,6 +50,7 @@ subtype "Metric"          => as "Maybe[CodeRef|Str]";
 subtype "ProfilerMode"    => as enum([qw[symbolic all]]);
 subtype "ProfilerState"   => as enum([qw[stop run]]);
 subtype "GradReq"         => as enum([qw[add write null]]);
+subtype "KVStoreStr"      => as enum([qw[local device dist dist_sync dist_async]]);
 subtype "NameShape"       => as "ArrayRef" => where {
     find_type_constraint("Str")->check($_->[0])
         and
@@ -57,6 +60,8 @@ subtype "Callback"        => as "CodeRef|ArrayRef[Coderef]|AI::MXNet::Callback|A
 subtype "EvalMetric"      => as "AI::MXNet::EvalMetric|Str|CodeRef";
 subtype "Optimizer"       => as "AI::MXNet::Optimizer|Str";
 subtype "Initializer"     => as "AI::MXNet::Initializer|Str";
+subtype "Updater"         => as "AI::MXNet::Updater|CodeRef";
+subtype "KVStore"         => as "AI::MXNet::KVStore|KVStoreStr";
 subtype "Activation"      => as "AI::MXNet::Symbol|Str|CodeRef";
 subtype "SymbolOrArrayOfSymbols" => as "AI::MXNet::Symbol|ArrayRef[AI::MXNet::Symbol]";
 subtype "NameShapeOrDataDesc" => as "NameShape|AI::MXNet::DataDesc";
```

tlby replied:
applied

bignamehyp committed:
The order of hue composed transformation (#7756)

* The order of hue composed transformation 

was incorrect. 
According to https://beesbuzz.biz/code/hsv_color_transforms.php
The final composed transform should be T_hsv  = T_rgb * T_H * T_S *  T_V * T_YIQ
translating to
        t = np.dot(np.dot(self.ityiq, bt), self.tyiq).T

With this correction, you can also change hue and saturation at the same time as  in https://beesbuzz.biz/code/hsv_color_transforms.php. 
For example:
        su = alpha_s * np.cos(alpha_h * np.pi)
        sw = alpha_s * np.sin(alpha_h * np.pi)
        bt = np.array([[alpha_v, 0.0, 0.0],
                       [0.0, su, -sw],
                       [0.0, sw, su]])
        t = np.dot(np.dot(self.ityiq, bt), self.tyiq).T
        src = nd.dot(src, nd.array(t))
        return src

* vsw/vsu means v*s*w/v*s*u

here only hue is changed, while v and s remain the same.

tlby commented:
I'm ready to squash these and create a PR for upstream, sound good?

sergeykolychev commented:
@tlby yes, sure the code looks good, see if the tests pass on master and file away!
